### PR TITLE
rate_limit: Add a global hook to rate limit concurrent connections based on SNI

### PR DESCRIPTION
**Pull request by zwoop**
There's still work to do here (see the README) such as proper, reloadable YAML configurations for the global hook. The remap version is likely also resource intensive when you have very large remap files.

Thanks to @shinrich for helping with the intricacies of the VC and SSL hooks and APIs.

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci autest]

File: plugins/experimental/rate_limit/sni_limiter.cc [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    break;
+  }
+
+  case TS_EVENT_VCONN_CLOSE: {
+    SniRateLimiter *limiter = static_cast<SniRateLimiter *>(TSUserArgGet(vc, gVCIdx));
+
+    if (limiter) {
+      limiter->release();
+    }
```
Comment (shinrich) [inline plugins/experimental/rate_limit/sni_limiter.cc:81]: Are you guaranteed that the vc is removed from the queue at this point?

**Review by shinrich — APPROVED** [review]
Looks good.  A nice generalization of the original mechanism.

**Comment by zwoop** [comment]
[approve ci clang-analyzer]

**Review by shinrich — DISMISSED** [review]

**Review by shinrich — APPROVED** [review]

**Comment by zwoop** [comment]
Cherry-picked to v9.1.x branch.

**Pull request merged**